packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -908,19 +925,33 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
           indexToLayoutOffset[index] = childParentData.layoutOffset!;
         }
 
+        // Initialize availableIndex to the current index to track the next free position if needed.
+        int availableIndex = index;
+
         if (newIndex != null && newIndex != index) {
           // The layout offset of the child being moved is no longer accurate.
           if (childParentData != null) {
             childParentData.layoutOffset = null;
           }
-
-          newChildren[newIndex] = _childElements[index];
+          if (!hasDuplicateKey(key)) {
+            // The child is being moved to a new index.
+            childParentData?.layoutOffset = null;
+
+            newChildren[newIndex] = _childElements[index];
+          } else {
+            // Handle duplicates: skip direct reordering, and add to the next available slot later.
+            // This finds the next free index if there's a duplicate key, preventing index conflicts.
+            while (newChildren.containsKey(availableIndex)) {
+              availableIndex++;
+            }
+            newChildren[availableIndex] = _childElements[index];
+          }
           if (_replaceMovedChildren) {
             // We need to make sure the original index gets processed.
             newChildren.putIfAbsent(index, () => null);
           }
-          // We do not want the remapped child to get deactivated during processElement.
-          _childElements.remove(index);
+          // Prevents the remapped child from being deactivated during processElement by using availableIndex.
```

Comment:
I don't immediately see a benefit to adding the `by using availableIndex` suffix. The line of code just below shows it's being used already?